include/ipc/ipc_service.h

```diff
@@ -170,6 +170,80 @@ int ipc_service_register_endpoint(const struct device *instance,
  */
 int ipc_service_send(struct ipc_ept *ept, const void *data, size_t len);
 
+/** @brief Get an empty TX buffer to be sent using @ref ipc_service_send_nocopy
+ *
+ *  This function can be called to get an empty TX buffer so that the
+ *  application can directly put its data into the sending buffer without copy
+ *  from an application buffer.
+ *
+ *  It is the application responsibility to correctly fill the allocated TX
+ *  buffer with data and passing correct parameters to @ref
+ *  ipc_service_send_nocopy function to perform data no-copy-send mechanism.
+ *
+ *  @param ept Registered endpoint by @ref ipc_service_register_endpoint.
+ *  @param data Pointer to the empty TX buffer.
+ *  @param len Pointer to store the TX buffer size.
+ *  @param wait Boolean, wait or not for buffer to become available
+ *
+ *  @retval -EIO when no backend is registered or send hook is missing from
+ *	    backend.
+ *  @retval -EINVAL when instance or endpoint is invalid.
+ *  @retval -ENOTSUP when this is not supported by backend.
+ *  @retval -ENOBUFS when there are no TX buffers available.
+ *
+ *  @retval 0 on success.
+ *  @retval other errno codes depending on the implementation of the backend.
+ */
+int ipc_service_get_buffer(struct ipc_ept *ept, void **data, uint32_t *len, int wait);
+
+/** @brief Send data in a TX buffer reserved by @ref ipc_service_get_buffer
+ *         using the given IPC endpoint.
+ *
+ *  This is equivalent to @ref ipc_service_send but in this case the TX buffer
+ *  has been obtained by using @ref ipc_service_get_buffer.
+ *
+ *  The application has to take the responsibility for getting the TX buffer
+ *  using @ref ipc_service_get_buffer and filling the TX buffer with the data.
+ *
+ *  After the @ref ipc_service_send_nocopy function is issued the TX buffer is
+ *  no more owned by the sending task and must not be touched anymore unless
+ *  the function fails and returns an error.
+ *
+ *  @param ept Registered endpoint by @ref ipc_service_register_endpoint.
+ *  @param data Pointer to the buffer to send obtained by @ref ipc_service_get_buffer.
+ *  @param len Number of bytes to send.
+ *
+ *  @retval -EIO when no backend is registered or send hook is missing from
+ *	    backend.
+ *  @retval -EINVAL when instance or endpoint is invalid.
+ *  @retval -EBADMSG when the message is invalid.
+ *  @retval -EBUSY when the instance is busy.
+ *
+ *  @retval 0 on success.
+ *  @retval other errno codes depending on the implementation of the backend.
+ */
+int ipc_service_send_nocopy(struct ipc_ept *ept, const void *data, size_t len);
+
+/** @brief Release the RX buffer for future reuse.
```

Comment:
I would split RX nocopy feature to other commit than TX nocopy (for clear separation).
For RX nocopy we need two functions like
* `rx_buffer_keep()`
* `rx_buffer_throw_away()`

To use zero-copy feature, user needs to call `rx_buffer_keep()` in `received()` callback implementation. When user stops using this buffer `rx_buffer_throw_away()` should be called.

If user does not call `rx_buffer_keep()` in `received()` callback implementation, typical copying operation is assumed and the buffer is automatically released after `received()` callback processing.

Reply:
maybe `claim` and `release` would be better.

Reply:
ok, I'll try to extend the RX path API.